### PR TITLE
resolve #94, #159, #292 and #299

### LINKED ISSUE #94: Self hosted Gitlab login provider support
Can this be done by just adding the URL of the self-hosted Gitlab to the configuration properties?

### LINKED ISSUE #159: Support arbitrary OpenID Connect providers for external authentication
# Feature request

I would like to be able to use arbitrary OpenID Connect providers for external authentication.

While this module has support for significantly more providers than the netlify fork, it is still limiting being stuck with the hardcoded providers listed here https://github.com/supabase/gotrue#external-authentication-providers

It looks like there is already a [Go OIDC client](https://github.com/coreos/go-oidc), so it should be possible to integrate relatively easily.

### PR DESCRIPTION
**Pull request by koba-ninkigumi**
## What kind of change does this PR introduce?

resolve this issue

#94 Self hosted Gitlab login provider support by exposing GOTRUE_EXTERNAL_GITLAB_URL as setting in UI
#159 Support arbitrary OpenID Connect providers for external authentication
#292 Allow org-specific endpoint to be specified for Azure OAuth provider
#299 Generic OpenID Connect auth provider

## What is the current behavior?

#94 Self hosted Gitlab login provider support by exposing GOTRUE_EXTERNAL_GITLAB_URL as setting in UI
#159 Support arbitrary OpenID Connect providers for external authentication
#292 Allow org-specific endpoint to be specified for Azure OAuth provider
#299 Generic OpenID Connect auth provider

## What is the new behavior?

Feel free to include screenshots if it includes visual changes.

## Additional context

https://github.com/supabase/supabase/pull/5181

### CODE REVIEW

**Comment by J0** [comment]
Looks good to me. We can merge this once #207 gets approved

File: api/token.go
Comment (kangmingtay) [inline api/token.go:353]: `getVerifierFromClientIDAndIssuer` seens to require both the `ClientID` and `Issuer` (https://github.com/supabase/gotrue/pull/360/files#diff-150d061ec4206a8396c1e267a3493c43049fb93cca6b50cb99cf4fa8c1adc2e7R107-R111), this check doesn't guarantee that the `ClientID` and `Issuer` will be present if the `Provider` is empty

Comment (kangmingtay) [reply api/token.go:353]: i think a clearer flow would be:

```
if params.Provider != "" {
  verifier, err = params.getVerifier(ctx)
} else if params.ClientID != "" && params.Issuer != "" {
  verifier, err = params.getVerifierFromClientIDAndIssuer(ctx)
} else {
  // return error here
}
```

Alternatively, we can also make this cleaner by having `getVerifier` be a wrapper around `getVerifierFromClientIDAndIssuer` and a new function called `getVerifierFromProvider` and do the check for the above inside `getVerifier`.

File: api/token.go
Comment (kangmingtay) [inline api/token.go:344]: shouldn't we need either the `Provider` or both the `ClientID` and `Issuer` ?

**Review by kangmingtay — CHANGES_REQUESTED** [review]

File: api/token.go
Comment (koba-ninkigumi) [reply api/token.go:344]: Clarification.

case 00  
  params.Provider == ""  
  params.ClientID == ""  
  params.Issuer == ""  
-> error

case 01  
params.Provider == ""  
params.ClientID == "abcde"  
params.Issuer == ""  
-> error

case 02  
params.Provider == ""  
params.ClientID == ""  
params.Issuer == "https://abcde"  
-> error

case 03  
params.Provider == "abcde"  
params.ClientID == ""  
params.Issuer == ""  
-> not error

case 04  
params.Provider == ""  
params.ClientID == "abcde". 
params.Issuer == "https://abcde"  
-> not error

case 05  
params.Provider == "abcde"  
params.ClientID == "abcde"  
params.Issuer == "https://abcde"  
-> not error

Comment (koba-ninkigumi) [reply api/token.go:344]: There is no need to change this part of the code, is there?

File: api/token.go
Comment (koba-ninkigumi) [reply api/token.go:352]: I followed your first advice.

**Comment by koba-ninkigumi** [comment]
@kangmingtay I followed your advice and completed the changes. 

Please review and approve the workflow.

File: api/token.go
Comment (kangmingtay) [reply api/token.go:344]: oh ok sry i misread it, yeah there's no need to change this

File: api/token.go
Comment (kangmingtay) [inline api/token.go:354]: I think we should return a `badRequestError` here since it's an issue with the request params being passed by the developer

**Review by kangmingtay — CHANGES_REQUESTED** [review]

File: api/token.go
Comment (koba-ninkigumi) [reply api/token.go]: @kangmingtay 
I have changed it. Are there any other problems?

**Review by kangmingtay — DISMISSED** [review]

**Comment by kangmingtay** [comment]
hey @koba-ninkigumi, before i merge this, can you provide an example where you would use a `client_id` and `issuer` with this endpoint? I want to test it out locally before merging this in thanks

**Comment by koba-ninkigumi** [comment]
@kangmingtay 
https://github.com/supabase/gotrue-js/pull/207
↑There is a sample at the end of my first post in this PR.
Replace the provider in this sample with the client_id and issuer, and run it.

**Comment by koba-ninkigumi** [comment]
@kangmingtay 
https://github.com/supabase/gotrue-js/pull/207
↑ This gotrue-js PR merge is also awaiting your approval.

**Comment by kangmingtay** [comment]
> @kangmingtay [supabase/gotrue-js#207](https://github.com/supabase/gotrue-js/pull/207) ↑ This gotrue-js PR merge is also awaiting your approval.

yup im leaving that to @dthyresson and @J0 to resolve, i'm gonna test this one out now and merge it if it's good.

**Comment by koba-ninkigumi** [comment]
@kangmingtay 

> yup im leaving that to @dthyresson and @J0 to resolve, i'm gonna test this one out now and merge it if it's good.

I understand.
However, 8 days ago, @dthyresson said the following.

> But defer to @kangmingtay and @J0 and @awalias and @thorwebdev.

I think your approval is also needed.
Please fulfill your responsibilities to each other.

**Review by kangmingtay — APPROVED** [review]

**Pull request merged**
